### DISCUSSION TITLE
pytest_sessionfinish tryfirst and trylast

### DISCUSSION THREAD

**Discussion by selenUser** [pytest-dev/pytest#10776] [Q&A]
Hi All

I run specific code in pytest_sessionfinish and tryfirst=true, all is well.
but now I want to use the created report and send it to a database, so I tried to use trylast=true, but some code not seeing in the report because it already 'close'.

How can I make this ?
into pytest_sessionfinish make some report changes, create the report xml file and then continue with more code ?

**Comment by portyu9** [comment]
Hi @selenUser 

A hook wrapper is a better fit than trying to register two separate `pytest_sessionfinish` implementations.

On current pytest:

```python
import pytest

@pytest.hookimpl(wrapper=True)
def pytest_sessionfinish(session, exitstatus):
    # Runs before the normal pytest_sessionfinish implementations.
    # Make any changes which must be included in the report here.

    yield

    # Runs after the normal implementations have finished.
    # At this point e.g. the JUnit XML plugin has finalized its output.
    send_report_to_database()
```

That gives you the ordering you described:

1. your pre-processing
2. pytest/plugin `pytest_sessionfinish` implementations, including report finalization
3. your post-processing

So if your second action needs to consume the **finished XML file**, put that action after `yield`.

Older pytest/pluggy versions used the older `hookwrapper=True` form, but for current pytest `wrapper=True` is the preferred form.

Please mark this answer as accepted if it helped, thank you.
